net/captivedetection/endpoints.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package captivedetection
+
+import (
+	"net/http"
+	"runtime"
+	"slices"
+
+	"tailscale.com/net/dnsfallback"
+)
+
+// availableEndpoints returns a set of Endpoints which can be used for captive portal detection by performing
+// one or more HTTP requests and looking at the response. The returned Endpoints are ordered by preference,
+// with the most preferred Endpoint being the first in the slice.
+func availableEndpoints(preferredDERPRegionID int) []Endpoint {
+	endpoints := []Endpoint{}
+
+	// Before anything, if we have a preferred DERP region, let's try to use that as our first endpoint.
+	preferredIPsAndHosts := []string{}
+	if preferredDERPRegionID != 0 {
+		hs := dnsfallback.GetDERPHostnames(preferredDERPRegionID)
+		ips := dnsfallback.GetDERPIPv4s(preferredDERPRegionID)
+		preferredIPsAndHosts = slices.Concat(ips, hs)
+	}
+
+	// First, try to use the DERP servers. Generate URLs for both their IPs and hostnames (IPs first).
+	// These might be blocked on some networks since they are IPs and hostnames associated with Tailscale.
+	derps := slices.Concat(preferredIPsAndHosts, dnsfallback.GetAllDERPIPv4s(), dnsfallback.GetAllDERPHostnames())
+	for _, derp := range derps {
+		url := "http://" + derp + "/generate_204"
+		e := Endpoint{url, http.StatusNoContent, "", true}
+		endpoints = append(endpoints, e)
+	}
+
+	// If all DERPs fail, let's try the default Tailscale coordination server. This is also blocked on some networks.
+	cs := Endpoint{"http://login.tailscale.com/generate_204", http.StatusNoContent, "", true}
+	endpoints = append(endpoints, cs)
+
+	// Lastly, to be safe, let's also include some well-known captive portal detection URLs that are not under the
+	// tailscale.com umbrella. These are less likely to be blocked on public networks since blocking them
+	// would break captive portal detection for many devices.
+	switch runtime.GOOS {
```

Comment:
Might be worth adding a test to make sure this function returns what we expect (both from the DERP map and here). One way we do that elsewhere is to take the `GOOS` as a parameter, so that we can test that this works on all platforms even if we're not running tests on that platform (e.g. iOS)

Reply:
done